### ISSUE TITLE
Build system / Client-side dependency management

### ISSUE THREAD

**Issue by mkalam-alami** [alakajam-team/alakajam#21]
If we run the server in production mode, Nunjucks should disable the style/script blocks and just use a minified version of CSS/JS files.

The build system should also apply the NextCSS filter to the site CSS.

Consider [Browserify](http://browserify.org/) (run build on startup in development)

In development, either serve directly from node_modules (how could we do that without duping script listing?), or use the browserify build as well.

**Comment by coxm** [comment]
One way to address static file URLs is to create a `static` nunjucks tag. We could either:
- pass names/IDs of static resources as and argument, e.g.
  `<script src="{% static "jquery_js" %}">`; or
- create a tag with two options, e.g.
  `<script src="{% static_prod %}/path/to/prod.min.js{% static_dev %}/path/to/dev.js{% endstatic %}">`.

In both cases we could do things like apply NextCSS (and maybe compile down from ES6 to ES5).

See https://mozilla.github.io/nunjucks/api.html#custom-tags

**Comment by ttencate** [comment]
Tentative plan, subject to change as I wrangle this:

[ ] Extract each inline `<script>`s to a separate file in a new top-level directory `client/`.
[ ] Make each Standard.js compliant.
[ ] Make each safe to include on any page without unwanted side effects.
[ ] Serve them all up as a single transpiled and minified bundle.

For making these scripts more "library-like", we could bind behaviour to elements by assigning a special CSS class. For example, the entry voting form would have `class="js-entry-vote"`. The `entry.js` script would have a DOM ready listener (`$(function () { ... })`), which checks for elements with that class and sets itself up on those. This should also ease any future transition to Knockout.js or some other data-binding library (#128).

**Comment by mkalam-alami** [comment]
Looks like a good plan. I just hope the code has not become too spaghetti-ish for easy refactoring, otherwise just doing this at the same time as #128 can be an alternative.

**Comment by coxm** [comment]
Regarding the polyfill, is there a particular reason not to use it? If those features are left out, can you enforce that they aren't used in the code with eslint rules? Otherwise we're likely to introduce difficult-to-find bugs.

**Comment by ttencate** [comment]
@coxm I share your concern, but decided like this for now because leaving out the polyfill doesn't make the current situation worse. I would think such bugs aren't very difficult to find: the console will say something along the lines of "foo.bar is not a function".

Also, of course, the polyfill would embiggen the bundle (103 kB minified, 34 kB gzipped) – unless you're aware of any good dead code elimination tools that work on vanilla JavaScript?

I'm all for a compile-time check, but I don't think eslint offers something like that (correct me if I'm wrong). It's possible that Flow or TypeScript can help us out here, or even provide compile-time transformations that make polyfills unnecessary.

**Comment by coxm** [comment]
Let's not get lost too far down an optimisation rabbit hole. For comparison I just loaded > 16MB of data from the homepage. :smile:

More worrying to me is that while these errors aren't hard to find for us developers, we will likely be using up-to-date browsers, so e.g. `Object.assign` *will* exist. On the other hand a user experiencing this with an older browser can't be relied upon to see console errors so we may not even find out.

Aside from that, these features are useful and I don't see why we should deprive ourselves... :stuck_out_tongue_winking_eye:

Can't we just use a polyfill served by a CDN, e.g. https://cdnjs.cloudflare.com/ajax/libs/babel-polyfill/6.26.0/polyfill.min.js ? If not, presumably the library will settle down eventually and we can set cache headers appropriately. In both cases I'd expect most users' browsers to have cached the result when we get to pinch points like entry submission. If neither of those are possible, but we're using ES6 modules, does this implementation provide tree shaking?

**Comment by ttencate** [comment]
I'm using `require()` for now because that's what Browserify understands. But I'm not opposed to switching to e.g. [rollup.js](https://rollupjs.org/). I don't know what I'm doing, after all :grinning:

**Comment by coxm** [comment]
In that case, can we just include the CDN polyfill linked above? Should be easy to insert it into a base template like `templates/_page.html`.

As for the PR, I've commented and approved. Not sure if @mkalam-alami wants to have a look so I'll leave open for now.

**Comment by ttencate** [comment]
OK, you've convinced me. The site is not using a CDN for any of its other third-party scripts either, so I'll just add babel-polyfill to be served up by Express for now. Possibly to be concatenated with other globally used libs later...

**Comment by ttencate** [comment]
I'm still alive! Going to push this through to the end.

It seems the world is [moving away](https://trends.google.com/trends/explore?date=today%205-y&q=browserify,webpack) from Browserify towards Webpack 2. It just seems more modern and feature-rich, although a little harder to get started with. Because we are still relatively uninvested at this point, I'll switch to Webpack if nobody objects. I want to stay away from using Webpack-isms like `require('style-loader!css-loader?modules!./styles.css')` though.

**Comment by ttencate** [comment]
This also ties in to autoreload during development. The default solution, `webpack-dev-middleware`, is problematic because it stores the bundle in memory only, so it would need to do a full rebuild every time server-side JS code is changed and `browser-refresh` restarts the server.

Instead I'll try an approach parallel to what we're doing for CSS: `postcss-watch` watches CSS input files and triggers a rebuild, while `browser-refresh` watches CSS output files and triggers a reload.

**Comment by ttencate** [comment]
I'm also going to bring some more consistency to the directory structure:

* `assets/` for input files (`assets/js/`, `assets/css/`, `assets/images/`), which need to be compiled somehow.
* `static/` for output files (`static/js/`, `static/css`, `static/images/`), which will be served up as-is.

This simplifies the logic of watching: builders like `postcess` watch `assets` and output to `static`, while `browser-refresh` never needs to watch `assets`, and knows that it can get away with a mere refresh (rather than server restart) if the file is in `static`.

**Comment by mkalam-alami** [comment]
Cool! I'll take a look at what it took to set up Webpack, indeed I've heard it can be harder to learn than some alternatives. Those JS/CSS building trends are getting hard to follow btw, techs are getting outdated before I have the time to master them :P

**Comment by mkalam-alami** [comment]
The current system looks good, created separate #243 to track the remaining code to migrate.

**Comment by ttencate** [comment]
Reopening this because the work on #243 made me think: what do we want to do with 3rd-party dependencies? For each of them, we have a choice:

1. **Host it ourselves from `/static/js/*.min.js`.** Pro: simple to drop in files. Con: not cached on first visit. Con: extra request. Con: does not tie in with the Webpack build system. Con: no tree shaking (dead code elimination).
2. **Pull it in from a CDN.** Pro: possibly cached even on first visit. Con: does not tie in with the Webpack build system. Con: no tree shaking.
3. **`npm install --save` it, and `require()` it from our own bundle.** Pro: fewer requests. Pro: tree shaking. Pro: more static analysis possible. Con: harder to see what else gets pulled in (bundle bloat).

I'm in favour of (3) for the lesser known stuff, and indeed I already did this for the smaller libraries (`sortablejs`, `flipclock` and `slug`).

```html
  <script type="text/javascript" src="/static/js/jquery.min.js"></script>
  <script type="text/javascript" src="/static/js/jquery.lazy.min.js"></script>
  <script type="text/javascript" src="/static/js/bootstrap.min.js"></script>
  <script type="text/javascript" src="/static/js/bootstrap-datetimepicker.min.js"></script>
  <script type="text/javascript" src="/static/js/icheck.min.js"></script>
  <script type="text/javascript" src="/static/js/select2.full.min.js"></script>
  <script type="text/javascript" src="/static/js/lodash.min.js"></script>
```

For `jquery`, `bootstrap` and `lodash` I'm not sure whether we should go for (3) or (2). An advantage of just bundling _everything_ ourselves is that we don't run the risk of pulling in e.g. jQuery twice (once as an indirect dependency in the bundle, once through CDN).

Thoughts? Opinions?

**Comment by mkalam-alami** [comment]
I noticed the issue as well, and ideally would favor (3), but I did not raised anything because of a (possibly invalid) performance concern with the Webpack build/server boot time.

I guess we could:

1. Do (3) fully
2. If the JS build time is significantly impacted, consider splitting the Webpack build into two output files (like `site.js` + `thirdparty.js`), with only `site.js` rebuilding during site development.
3. *(Bonus)* Consider adding a plugin like [only-if-changed](https://www.npmjs.com/package/only-if-changed-webpack-plugin) for faster build/boot times in general

**Comment by ttencate** [comment]
Webpack build times are in the 500-600 ms range for me during small changes, and 1700 ms at startup (development mode), so thus far it's acceptable. I agree we can just go all in, and fix any performance issues as they crop up.

There seems to be some change detection already, because a no-op save rebuilds in 60 ms instead of 500 ms. I briefly tried `only-if-changed` and if anything, it made things slower.

Following https://webpack.js.org/guides/build-performance/, I upgraded Webpack from 3.8.1 to 4.3.0 ([4.0.0 release notes](https://github.com/webpack/webpack/releases/tag/v4.0.0) have a list of breaking changes). It seems that the `cheap-module-eval-source-map` option makes a big difference, bringing build times for actually changed files down to around 200 ms. Not sure if this actually requires 4.3.0 but whatever, it works now.

A possible future extension is the `cache-loader`, which caches compilation results on disk.

**Comment by mkalam-alami** [comment]
FYI I've continued experimenting on the topic over at https://github.com/alakajam-team/alakajam-poc

On that repo, the main changes are:
- Switched to TypeScript for the client-side sources
- Started to pull 3rd party libs (Bootstrap, jQuery, Vue.js _(!)_ ) from `node_modules` into the TypeScript build
- Used Webpack to also package our CSS (with postcss like before)
- Included the 3rd party CSS (Bootstrap, Font Awesome) into our build (a bit tricky, especially with external resources like images & fonts)

For me the TypeScript build already takes 6.5s, and the CSS one 5.5s :disappointed: So not sure if it's worth bothering with integrating the 3rd party libs that way.

**Comment by mkalam-alami** [comment]
To be continued at #398
